### PR TITLE
feat(system-server): add GET /system/connected

### PR DESCRIPTION
**Pull request by fsinapi**
# Overview

This PR adds in a route `GET /system/connected` that will return information about all of the currently connected clients. In order to implement this, an `AuthorizationTracker` class keeps track of the clients that obtain a token through the `POST /system/authorize` route.

Closes RCORE-520 and RCORE-515

Requires #12198, will change the base branch to `edge` once that is merged

# Test Plan

Tested on a Dev server & through integration testing:

- Register a client, then ping /system/connected and ensure _nothing_ pops up (registration does not count as connecting)
- Authorize that client, ping /system/connected, ensure that the correct Registrant info shows up
- Authorize the client again, ping /system/connected, ensure the same info comes back (no duplication)
- Register & authorize a second client, ping /system/connected, ensure _both_ sets of Registration info come back correctly

# Changelog

- Added `AuthorizationTracker` class that maintains a list of authorizations & their expiration datetimes
- Added `expiration_from_jwt` function to extract the expiration time from a JWT
- Added functionality in `POST /system/authorize` to add authorizations to a singleton tracker
- Added `GET /system/connected` route that extracts authorizations from the singleton tracker
- Modified the dev server fixture to run a new instance for each test - this increases test time a bit, but establishes a clean server state after each test.

# Review requests

# Risk assessment

Low

### CODE REVIEW

File: system-server/system_server/connection/authorization_tracker.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        Args:
+            registrant: Information about the registrant for this connection
+
+            expiration: When this registrant is considered expired.
+        """
+        # If this registrant already exists, remove it
+        for c in self._connections:
+            if c.registrant == registrant:
+                self._connections.remove(c)
```
Comment (sfoster1) [inline system-server/system_server/connection/authorization_tracker.py]: this alters `self._connections` while iterating through it which is not a good idea. Maybe put the found value in your pocket and remove it after the end of the iteration, or do a filtered list copy

File: system-server/system_server/connection/authorization_tracker.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        self._connections.append(
+            _Authorization(registrant=registrant, expiration=expiration)
+        )
+
+    def _update_active_connections(self) -> None:
+        """Clear out any expired connections."""
+        now = datetime.now()
+        for c in self._connections:
+            if c.expiration < now:
+                self._connections.remove(c)
```
Comment (sfoster1) [inline system-server/system_server/connection/authorization_tracker.py]: same deal - this alters `_connections` while iterating through it, may be better as `self._connections = [c for c in self._connections if c.expiration > now]`

File: system-server/system_server/jwt/system_jwt.py [hunk 3]
```diff
@@ -38,6 +38,19 @@ def registrant_from_jwt(token: str, signing_key: str) -> Registrant:
...
+def expiration_from_jwt(token: str, signing_key: str) -> datetime:
+    """Decode the expiration time from a JWT."""
+    decoded = jwt.decode(
+        jwt=token,
+        key=signing_key,
+        algorithms=[_JWT_ALGORITHM],
+        options={"verify_aud": False},
+    )
+    exp = decoded["exp"]
+    print(f"Timestamp: {exp}")
```
Comment (sfoster1) [inline system-server/system_server/jwt/system_jwt.py]: should be a `log` or removed

File: system-server/system_server/jwt/system_jwt.py [hunk 4]
```diff
@@ -38,6 +38,19 @@ def registrant_from_jwt(token: str, signing_key: str) -> Registrant:
...
+    return datetime.fromtimestamp(float(exp))
```
Comment (sfoster1) [inline system-server/system_server/jwt/system_jwt.py]: `float` doesn't seem right here, shouldn't this be an int?

File: system-server/system_server/persistence/__init__.py [hunk 5]
```diff
@@ -29,6 +33,7 @@
 _persistence_dir_lock = Lock()
 _sql_lock = Lock()
 _uuid_lock = Lock()
+_authorization_tracker_lock = Lock()
```
Comment (sfoster1) [inline system-server/system_server/persistence/__init__.py:37]: at some point we should refactor so these locks are held in application state

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Structure looks great, couple nits around altering collections while iterating through them

File: system-server/system_server/persistence/__init__.py (hunk 5 above)
Comment (SyntaxColoring) [reply system-server/system_server/persistence/__init__.py:37]: Yep, this has been bothering me in `robot-server` too. It's a good candidate for some kind of reusable util.

**Review by sfoster1 — APPROVED** [review]
Looks good to me!

**Pull request merged**